[mlir] Added new attributes to the llvm.call op in llvmir target

FMarno opened the pull request:
The new attributes are:
 * convergent
 * no_unwind
 * will_return
 * memory effects

FMarno committed:
[mlir] Added new attributes to the llvm.call op in llvmir target

The new attributes are:
 * convergent
 * no_unwind
 * will_return
 * memory effects

Dinistro commented on mlir/lib/Target/LLVMIR/ModuleImport.cpp:
```diff
@@ -1471,6 +1471,30 @@ LogicalResult ModuleImport::convertInstruction(llvm::Instruction *inst) {
+
+    // memory effects
+    llvm::MemoryEffects memEffects = callInst->getMemoryEffects();
+    // Only set the attr when it does not match the default value.
+    auto othermem = convertModRefInfoFromLLVM(
+        memEffects.getModRef(llvm::MemoryEffects::Location::Other));
+    auto argMem = convertModRefInfoFromLLVM(
+        memEffects.getModRef(llvm::MemoryEffects::Location::ArgMem));
+    auto inaccessibleMem = convertModRefInfoFromLLVM(
+        memEffects.getModRef(llvm::MemoryEffects::Location::InaccessibleMem));
+    auto memAttr = MemoryEffectsAttr::get(callOp.getContext(), othermem, argMem,
+                                          inaccessibleMem);
+    if (!memAttr.isReadWrite()) {
+      callOp.setMemoryAttr(memAttr);
+    }
```
```suggestion
    auto othermem = convertModRefInfoFromLLVM(
        memEffects.getModRef(llvm::MemoryEffects::Location::Other));
    auto argMem = convertModRefInfoFromLLVM(
        memEffects.getModRef(llvm::MemoryEffects::Location::ArgMem));
    auto inaccessibleMem = convertModRefInfoFromLLVM(
        memEffects.getModRef(llvm::MemoryEffects::Location::InaccessibleMem));
    auto memAttr = MemoryEffectsAttr::get(callOp.getContext(), othermem, argMem,
                                          inaccessibleMem);
    // Only set the attr when it does not match the default value.
    if (!memAttr.isReadWrite())
      callOp.setMemoryAttr(memAttr);
```

FMarno replied:
That makes sense. addressed.

Dinistro commented on mlir/lib/Target/LLVMIR/ModuleImport.cpp:
```diff
@@ -1471,6 +1471,30 @@ LogicalResult ModuleImport::convertInstruction(llvm::Instruction *inst) {
+    // memory effects
+    llvm::MemoryEffects memEffects = callInst->getMemoryEffects();
+    // Only set the attr when it does not match the default value.
+    auto othermem = convertModRefInfoFromLLVM(
```
Nit: Don't use `auto` here, as the type is not given on the RHS.

FMarno replied:
Sure, done

Dinistro commented on mlir/test/Dialect/LLVMIR/call.mlir:
Don't we already check other call attributes in another test? Would make sense to keep them all together.

FMarno replied:
I've moved the tests as suggested

Dinistro commented on mlir/test/Target/LLVMIR/Import/call-attributes.ll:
As above

Dinistro reviewed (commented):
Thanks for working on this, dropped a few nit comments.

victor-eds commented on mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp:
```diff
@@ -219,6 +219,28 @@ convertOperationImpl(Operation &opInst, llvm::IRBuilderBase &builder,
+      call->addFnAttr(llvm::Attribute::WillReturn);
+
+    // memory effects
+    if (callOp.getMemory()) {
+      MemoryEffectsAttr memAttr = callOp.getMemoryAttr();
```
```suggestion
    if (MemoryEffectsAttr memAttr = callOp.getMemoryAttr()) {
```

This should work

FMarno replied:
yep, done

victor-eds commented on mlir/test/Target/LLVMIR/Import/call-attributes.ll:
```diff
@@ -0,0 +1,53 @@
+define void @call_will_return() {
+; CHECK: llvm.call @f() {memory = #llvm.memory_effects<other = none, argMem = none, inaccessibleMem = none>}
+  call void @f() memory(none)
+; CHECK: llvm.call @f() {memory = #llvm.memory_effects<other = none, argMem = write, inaccessibleMem = read>}
+  call void @f() memory(none, inaccessiblemem: read, argmem: write)
```
Really a nit, but can we sort?

```suggestion
; CHECK: llvm.call @f() {memory = #llvm.memory_effects<other = none, argMem = write, inaccessibleMem = read>}
  call void @f() memory(none, argmem: write, inaccessiblemem: read)
```

victor-eds commented:
Not really to be addressed in this PR, but IMO this and the `llvm.func` counterpart attribute should be called `memory_effects` as it is the case in LLVM IR. @ftynse are you aware of why the LLVM dialect decided to go with this name instead?

FMarno committed:
amend! small style changes + new test

[mlir] Added new attributes to the llvm.call op in llvmir target

The new attributes are:
 * convergent
 * no_unwind
 * will_return
 * memory effects

Dinistro commented:
> Not really to be addressed in this PR, but IMO this and the `llvm.func` counterpart attribute should be called `memory_effects` as it is the case in LLVM IR. @ftynse are you aware of why the LLVM dialect decided to go with this name instead?

I introduced this a while ago, there is no reason to why it's named like this. Feel free to rename for consistency

FMarno commented:
> > Not really to be addressed in this PR, but IMO this and the `llvm.func` counterpart attribute should be called `memory_effects` as it is the case in LLVM IR. @ftynse are you aware of why the LLVM dialect decided to go with this name instead?
> 
> I introduced this a while ago, there is no reason to why it's named like this. Feel free to rename for consistency

I'd be happy to do this change, but I would rather do it in another PR to minimize the changes here.

FMarno committed:
amend! move tests

[mlir] Added new attributes to the llvm.call op in llvmir target

The new attributes are:
 * convergent
 * no_unwind
 * will_return
 * memory effects

FMarno commented on mlir/test/Dialect/LLVMIR/call.mlir:
```diff
@@ -0,0 +1,34 @@
+// RUN: mlir-opt -split-input-file -verify-diagnostics %s | mlir-opt | FileCheck %s
+
+module {
```
thanks for the tip, done

Dinistro reviewed (commented):
Thanks for addressing the comments, LGTM!
Side note: We should really attempt to cleanup some of these test files in the future, as the LLVM dialect tests are a huge mess. I might find some time at some point to do so.

gysit commented on mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp:
```diff
@@ -219,6 +219,27 @@ convertOperationImpl(Operation &opInst, llvm::IRBuilderBase &builder,
+
+    // memory effects
+    if (MemoryEffectsAttr memAttr = callOp.getMemoryAttr()) {
+      llvm::MemoryEffects newMemEffects =
```
```suggestion
      llvm::MemoryEffects memEffects =
```
nit: If possible I would avoid new here.

gysit reviewed (approved):
Thanks!

LGTM modulo some ultra nits.

FMarno committed:
amend! style fixes

[mlir] Added new attributes to the llvm.call op in llvmir target

The new attributes are:
 * convergent
 * no_unwind
 * will_return
 * memory effects

FMarno commented on mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp:
```diff
@@ -219,6 +219,27 @@ convertOperationImpl(Operation &opInst, llvm::IRBuilderBase &builder,
+    if (callOp.getWillReturnAttr())
+      call->addFnAttr(llvm::Attribute::WillReturn);
+
+    // memory effects
```
Thanks for your suggestions, I've included them all

FMarno committed:
amend! format

[mlir] Added new attributes to the llvm.call op in llvmir target

The new attributes are:
 * convergent
 * no_unwind
 * will_return
 * memory effects
